### DISCUSSION TITLE
(feature request) Aggregate metrics from worker nodes: Histogram instead of percentiles

### DISCUSSION THREAD

**Discussion by bdarlt** [artilleryio/artillery#1644] [Artillery]
This is half a feature request, and half "Would I be nuts to implement this?"

**Use case**:
When running artillery in Kubernetes or similar configuration, ideally one wants to be able to accurately aggregate the metrics from each node into a test wide whole.

This is simple enough for counters (I haven't looked at rates), but difficult with percentiles.

**The problem**:
Artillery's summary/histogram metric uses percentiles, which are [troublesome to aggregate](https://www.circonus.com/2018/11/the-problem-with-percentiles-aggregation-brings-aggravation/).  Loss of accuracy can happen around the extremes, and can lead to misleading results.

**Proposed Solution**:
If a histogram could be used instead, the information is easily merged, you sum each bucket together.  Percentiles can be [calculated using the buckets](https://sciencing.com/calculate-cumulative-relative-frequency-8595620.html).  The buckets act to define a range of values, and the value in the bucket is a counter of how many things are in the range.  Thus one can sum the histogram data just like an array of counters -- because it is an array of counter.

Statsd (Datadog and Telegraf too) and Prometheus have support for histogram datatypes.  For local or single node reporting a calculation can be made to continue using percentiles somewhat transparently inside the "report" generation functions.  And for remote metric systems, the change would be to use the histogram data directly.

**Critisism of solutions**
Histograms aren't really meant to calculate percentiles.  If accurate percentiles are desired for single node scenarios, one could calculate as before in addition to the histogram.

However, using a log-linear histogram (so you don't have to calculate the buckets) seems to yield a fine enough bucket size that inaccuracies from calculating from buckets to percentiles would hopefully be minimal.

**Comment by hassy** [comment]
Hi @bdarlt 👋 Not nuts at all, and it's something Artillery already supports. :) 

If you run a [distributed load test on AWS Lambda](https://www.artillery.io/docs/guides/guides/distributed-load-tests-on-aws-lambda) with Artillery today, metrics are aggregated and merged from all workers and presented as one report. (Likewise, if you run a distributed test on Fargate with Artillery Pro.)

Artillery's percentile reporting is powered by [DDSketch](https://www.datadoghq.com/blog/engineering/computing-accurate-percentiles-with-ddsketch/) which is a high-precision fully-mergeable histogram implementation. Percentile values reported by Artillery are guaranteed to be accurate regardless of the number of Artillery nodes (and the distribution of the number of requests between nodes).

**Reply by bdarlt** [reply]
Hmm... re-reading the code I understand most of that now.  Thank you for your reply.

I'm trying to run Artillery much as you are in AWS Lambda (or Fargate).  But in Azure Container Instances.  So I'm using the local platform and orchastrating the containers myself.  I'd love a kubernetes operator... but I'm trying to get this bit working first.

I've been collecting status via prometheus, and am seeing median, min, max, p99 etc.  But I don't want to. Prometheus supports histograms.

However, I'm happy to switch to telegraf/influxdb line protocol (I've been wanting to run VictoriaMetrics for several reasons, this is just one more).  I just didn't see any reason they would be different.

**Reply by hassy** [reply]
Azure Container Instances look very similar to Fargate. It should *in theory* be possible to add first-class support for ACI as a compute platform into Artillery, similarly to what we're doing for Lambda & Fargate. That layer of the code is written with allowing exactly that in mind, although there are still plenty of rough edges. If that's something you might be interested in looking at let me know, I'd be happy to help out!

**Reply by bdarlt** [reply]
Perhaps... My hard requirement is that playwright support is present.

As it stands, I deploy the single node version of artillery on Z nodes, and let it rip - this has a few draw backs - you can't orcastrate load very well, dividing test data amung the nodes is left to each test, and stats don't aggrate (what started this thread).  The plus, however, is all Artillery functionality is present.

I've read through the code quickly, but don't know AWS enough to say how well what you have maps to Azure.  
From what I can tell, you spawn worker nodes, and use Simple Queue Service to handle communications.  I

Very roughly speaking, I had thought of doing the JMeter aproach - N worker nodes, controlled by a controller.  Replacing RMI with an messaging system - you use SQS, Azure Queue Storage is the equivalent.  For now, I'd by-pass any dynamic node creation (asking for one more, or one less) and go with a static number of workers decided at launch of the controller.  It could be added later using an Azure Funtion that creates/destroys as needed.

I worry about the amount of effort though.  The use of SQS (a platform dependancy) doesn't appear to have been abstracted.
